### PR TITLE
chore/babel

### PR DESCRIPTION
**Pull request by xaverfleer**
This addresses issue #14. Babel allows to use next generation JavaScript. E.g. Babel compiles arrow functions to IE11 compatible functions.

PR is not ready to merge back. Open tasks
- [x] fix netlify build command
- [x] adapt eslint
- [x] go through review entries
- [x] find solution for compiling / outsourcing javascript in script tags -> @xaverfleer found no way to compile JavaScript in HTML tags with Babel. It is probably good practice to isolate JavaScript code into .js files.

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -40,8 +40,6 @@ module.exports = {
     // disable shorthand property names because IE11 lacks support: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#Browser_compatibility
     'object-shorthand': ['error', 'never',],
     // disable arrow functions because IE11 lacks support: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#Browser_compatibility
-    'prefer-arrow-callback': 0,
-    // allow the use of console for server-side JS
```
Comment (raphaelschaad) [inline .eslintrc.js]: Why are you deleting this comment? I think it's a slip and you wanted to delete the comment above.

File: package.json [hunk 2]
```diff
@@ -11,7 +11,8 @@
     "start": "npm-run-all start-*",
     "start-lambda": "netlify-lambda serve ./lambda-functions/",
     "build": "npm-run-all build-*",
-    "build-lambda": "netlify-lambda build ./lambda-functions/"
+    "build-lambda": "netlify-lambda build ./lambda-functions/",
+    "build-babel": "babel src -d public"
```
Comment (raphaelschaad) [inline package.json]: `./public/` for local consistency

File: .eslintrc.js (hunk 1 above)
Comment (xaverfleer) [reply .eslintrc.js]: you are correct

File: package.json [hunk 3]
```diff
@@ -29,6 +30,9 @@
     "twilio": "^3.29.2"
   },
   "devDependencies": {
```
Comment (raphaelschaad) [inline package.json]: So, effectively, Babel is already in our project, brought in as an indirect dependency. I assume npm handles this gracefully and Babel now isn't "added twice" to but just as a direct dependency?

File: README.md [hunk 4]
```diff
@@ -33,7 +33,9 @@ To exclude sections of code from linting (e.g. copied and minified 3rd party sni
 Run `$ npm run start` (shorthand `$ npm start` to start a web server. We currently have `$ npm run start-lambda` that starts the [netlify-lambda package](https://www.npmjs.com/package/netlify-lambda) (a simple Webpack/Babel setup) to test lambda functions locally. It watches `./lambda-functions/`, builds any contained functions (including transpiled JS/TS features and imported modules) into `./lambda-functions-dist/`, and serves them to <http://localhost:9000/functionname>. Netlify Functions currently (early 2019) run by default in Node v8.10 and you may need to run the same version (`$ nvm install 8.10`, to switch back `$ nvm use stable`) to mirror this environment locally.
 
 ### Run building
-Run `$ npm run build` to create a distribution of the project for production. We currently have `$ npm run build-lambda` that we include in the `netlify.toml` to run when deploying.
+Run `$ npm run build` to create a distribution of the project for production. These commands get triggered:
+- `$ npm run build-lambda` builds `lambda-functions-dist` for Netlify's AWS Lambda functions.
+- `$ npm run build-babel` builds `public/*.js` for the HTML pages.
```
Comment (raphaelschaad) [inline README.md]: `./public/*.js` for local consistency (e.g. the section above).

So -- is this comment still true when looking at the src/ directory?

File: .gitignore [hunk 5]
```diff
@@ -3,3 +3,4 @@
 .env
 /node_modules/
 /lambda-functions-dist/
+/public/*.js
```
Comment (raphaelschaad) [inline .gitignore]: Is adding a gitignore like this common practice to avoid JS files in the public directory?

Separately, do you suggest the public directory should be named public or dist, now that it will not only hold public static files but "built" JS from a src directory?

**Comment by raphaelschaad** [comment]
Nice, I've completed a first pass.

File: .eslintrc.js (hunk 1 above)
Comment (xaverfleer) [reply .eslintrc.js]: 830066c fixes the comment

File: package.json [hunk 6]
```diff
@@ -11,7 +11,7 @@
     "start": "npm-run-all start-*",
     "start-lambda": "netlify-lambda serve ./lambda-functions/",
     "build": "npm-run-all build-*",
-    "build-babel": "babel src -d public",
+    "build-babel": "babel ./src -d ./public",
```
Comment (raphaelschaad) [inline package.json]: All the folders above have trailing slashes. In [my comment, I also included trailing slashes](https://github.com/raphaelschaad/Plus-Web/pull/26#discussion_r269806079). A convention is that "A directory path is only distinguishable from a file path if the directory path has a trailing slash." I don't feel super passionate about this particular convention -- it will work either way. But I *do* feel strongly about local consistency and about accurate addressing of feedback.

**Comment by xaverfleer** [comment]
Thank you for the review, great inputs! I will dig a little deeper into the babel docs before tackeling the details.

File: README.md [hunk 7]
```diff
@@ -56,6 +56,11 @@ For any private/"behind a login" pages (even though some browsers use the meta d
 
 A good meta description is unique for each page, in its respective language, ~150 characters long, optimized to provide value, and drive click-through in the first few words.
 
+## Code conventions
```
Comment (raphaelschaad) [inline README.md]: Have you seen the `## Code guidelines` above?

File: package.json (hunk 2 above)
Comment (xaverfleer) [reply package.json]: fixed links

I did not get the question. Are you asking about the internals of babel? Or are you asking how we can automate the compiling step?

File: package.json (hunk 3 above)
Comment (xaverfleer) [reply package.json]: that's right, as long as the versions match, npm can use the same download

File: README.md (hunk 4 above)
Comment (xaverfleer) [reply README.md]: 296710c fixes `./public/*.js` for local consistency

Comment (xaverfleer) [reply README.md]: 3f736a1 improves comment

File: .gitignore (hunk 5 above)
Comment (xaverfleer) [reply .gitignore]: > Is adding a gitignore like this common practice to avoid JS files in the public directory?

not sure. you know probably more projects. At Bucher + Suter we only version source files since the compile step is automatic and can be reproduced whenever necessary.

> Separately, do you suggest the public directory should be named public or dist, now that it will not only hold public static files but "built" JS from a src directory?

No strong opinion. To me both names are synonyms in this context (dist = public = distributed to the public). Do you have a preference?

File: package.json (hunk 6 above)
Comment (xaverfleer) [reply package.json]: > about accurate addressing of feedback.

😔. 🎯🎯🎯

> I don't feel super passionate about this particular convention

this works for me. 1df2974 adds rules for the project.

> local consistency

just to be sure: is "local" the file or the whole project?

File: README.md (hunk 7 above)
Comment (xaverfleer) [reply README.md]: thank you. 7dfa46f moves the section

File: package.json (hunk 2 above)
Comment (raphaelschaad) [reply package.json]: My question was poorly written. Is it correct that babel will take everything from src/, transpile it, and output it under the same filename into public/?

File: .gitignore (hunk 5 above)
Comment (raphaelschaad) [reply .gitignore]: Ignoring generated files seems like a good practice. What seems a bit confusing is to exclude all JS files from a folder named public. What if someone comes along and writes a JS (source) file and puts it in public. So my preference is to rename public/ to dist/. That way it's a bit clearer and mirrors lambda-functions-dist/. It probably would make sense to move static files out of dist/ and into a static/ folder. But we don't need to do this as part of this PR.

File: package.json (hunk 6 above)
Comment (raphaelschaad) [reply package.json]: "It depends." Local usually means the current scope that makes sense for the context. Usually, file. But you could say:
- block
- function
- class
- file
- module
...

For example, a class could be using an older language syntax for functions. When adding a new function, you would add it like the other functions. That way it's locally consistent. And then you can update them to the new syntax when it makes sense separately.

File: README.md (hunk 7 above)
Comment (raphaelschaad) [reply README.md]: Have you seen it? :)

The change looks good!

File: package.json (hunk 2 above)
Comment (xaverfleer) [reply package.json]: > Is it correct that babel will take everything from src/, transpile it, and output it under the same filename into public/?

correct

File: .gitignore (hunk 5 above)
Comment (xaverfleer) [reply .gitignore]: I am not sure if I understand this correctly. I try:

- `/public/` contains all files accessible from the website (generated and written ones).
- `*-dist/` contains generated files (public and private ones).
- What you propose is to compile from `/src/` to `/public/dist/`.

Is this correct?

File: package.json (hunk 6 above)
Comment (xaverfleer) [reply package.json]: makes sense, thank you

File: README.md (hunk 7 above)
Comment (xaverfleer) [reply README.md]: I had seed it, but did not have it present in my mind.

File: .gitignore (hunk 5 above)
Comment (raphaelschaad) [reply .gitignore]: Bullet one and two is correct. Bullet three -- I think -- could work like you said isn't needed and might complicate things. See my altered comment above. Maybe a comment in the .gitignore above the line in question is the the most straight-forward answer: `# Ignore generated files`

Comment (xaverfleer) [reply .gitignore]: 3cba607 adds the comment.

File: src/main.js [hunk 8]
```diff
@@ -24,7 +24,7 @@ if (typeof window !== 'undefined') {
   // Dynamic lang selector (DE/EN) for 404 page not found
   const langSelectorEl = document.querySelector('#lang-selector');
   if (langSelectorEl) {
-    if (window.location.pathname.indexOf('/en/') === 0) {
+    if (window.location.pathname.startsWith('/en/')) {
```
Comment (raphaelschaad) [inline src/main.js]: This change (semantically) shouldn't be part of this commit. No need to fix it tho.

File: README.md [hunk 9]
```diff
@@ -37,7 +37,11 @@ Run `$ npm run lint` to run linting. We currently have `$ npm run lint-js` to li
 Run `$ npm run start` (shorthand `$ npm start` to start a web server. We currently have `$ npm run start-lambda` that starts the [netlify-lambda package](https://www.npmjs.com/package/netlify-lambda) (a simple Webpack/Babel setup) to test lambda functions locally. It watches `./lambda-functions/`, builds any contained functions (including transpiled JS/TS features and imported modules) into `./lambda-functions-dist/`, and serves them to <http://localhost:9000/functionname>. Netlify Functions currently (early 2019) run by default in Node v8.10 and you may need to run the same version (`$ nvm install 8.10`, to switch back `$ nvm use stable`) to mirror this environment locally.
 
 ### Run building
```
Comment (raphaelschaad) [inline README.md]: Very nice change.

**Comment by xaverfleer** [comment]
> What’s the reasoning behind removing JS linting within HTML files in 2b187b5 ? 

Short answer: This is an error
Longer answer: We introduce Babel among other things to use ES2015+ syntax like arrow functions - unfortunately Babel only supports compiling of `.js` files. This means we need an ESLint configuration (`.eslintr.js`) for `.js` files another (`.eslintrcHtml.js`) for `<script>` tags in HTML files. de51d55 adds this second configuration.

> (technically not “compile”, by the way)

I try to stick to "official" definitions, in this case to [babeljs.io](https://babeljs.io/) that presents itself as "Babel is a JavaScript compiler"

File: README.md [hunk 10]
```diff
@@ -26,15 +31,17 @@ For local testing of backend functions, define environment variables for sensiti
 Run `$ npm run test` (shorthand `$ npm test`) to run tests. We currently have `$ npm run test-unit` to run the unit tests in `mainTest.js`. We use [tape](https://github.com/substack/tape) as our light-weight testing framework.
 
 ### Run linting
-Run `$ npm run lint` to run linting. We currently have `$ npm run lint-js` to lint all JS in `*.js` and `*.html` files according to the rules in `.eslintrc.js` (ignore files and folders with `.eslintignore`). Our current base choice are the `airbnb-base` rules (alternatives would be `eslint-config-google` and `standard`). We use [ESLint](https://eslint.org/), a linting utility for JS/JSX but not JSON/HTML/XML.
-
-To exclude sections of code from linting (e.g. copied and minified 3rd party snippets), enclose it by `/* eslint-disable */` ... `/* eslint-enable */` as [documented here](https://eslint.org/docs/user-guide/configuring#disabling-rules-with-inline-comments).
```
Comment (raphaelschaad) [inline README.md]: Should this be brought back?

**Comment by raphaelschaad** [comment]
You're right, compiler is also right. Transpiler is also right and more specific. While I usually think the more specific the better, this distinction doesn't seem worthwhile. The Babel guys probably thought the same.

File: README.md (hunk 10 above)
Comment (xaverfleer) [reply README.md]: Your have sharp eyes: it should

I just re-implemented the branch with cleaner commits. The section is back in ecab3f1.

File: .eslintrcHtml.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+      {
+        // adds exception for variable that already existed before eslint was introduced
+        allow: ['_wq'],
+      },
+    ],
+    // disable shorthand property names because IE11 lacks support: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#Browser_compatibility
+    'object-shorthand': ['error', 'never'],
+    // disable arrow functions because IE11 lacks support: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#Browser_compatibility
+    'prefer-arrow-callback': 0,
+    // allow the use of console for server-side JavaScript
```
Comment (raphaelschaad) [inline .eslintrcHtml.js]: This can be omitted -- shipping console statements in HTML seems like a code smell.

**Comment by raphaelschaad** [comment]
Given browserlists comes as a dependency with babel, would it make sense to codify our meager README comment on supported browsers [like described on the babel website](https://babeljs.io/docs/en/babel-preset-env#browserslist-integration) as part of this PR? If there's a good reason not to, I'm happy to drop it.

File: .eslintrcHtml.js (hunk 11 above)
Comment (xaverfleer) [reply .eslintrcHtml.js]: I agree. 4507e49 removes the rule.

What is the reason, why we need `console.log` for server-side JavaScript? Is it for debugging only?

**Comment by xaverfleer** [comment]
> would it make sense to codify our meager README comment on supported browsers

Could you specify, what you mean by "codify"? Do you mean replace the README.md comment with a new comment like `{
  "targets": {
    "chrome": "58",
    "ie": "11"
  }
}`?

**Comment by xaverfleer** [comment]
> would it make sense to codify our meager README comment on supported browsers

I got it. 25dd37d improves the comment.

File: .eslintrcHtml.js (hunk 11 above)
Comment (raphaelschaad) [reply .eslintrcHtml.js]: 👍

Server-side logging is a good practice. You always want to log everything on the server and keep those logs around to monitor, analyze, inspect, debug. For that, usually robust logging frameworks are used. Currently, I'm doing a poor man's job with JS console.logs over which I don't even have full control, I can just temporarily see them when logging into Netlify. I currently review them every time the MVP gets used "productively".

**Comment by raphaelschaad** [comment]
Yes! But ideally not replace the text in the README but enforce it with code (browserlists).

**Comment by raphaelschaad** [comment]
FYI master has moved. Watch especially out for https://github.com/raphaelschaad/Plus-Web/commit/1198bea3231bf4e86e8b3c3d80dcfd243071acb9 which adds an entire folder of HTML files that were copied from a pre-this-PR state (and might require updating as part of this PR).

**Comment by raphaelschaad** [comment]
From your side, are we ready to merge this?

**Comment by xaverfleer** [comment]
I rebased chore/babel to the master.
I added Polyfill to the new HTML documents.
I regrouped the commits in units that make sense.

Now, we are ready to merge.

**Comment by raphaelschaad** [comment]
LGTM! While playing around with it I've noticed a syntax error in a comment and fixed it. It seems IE11 is included because we need that one (I've also pushed a tiny wording tweak to clarify that in the README).

Before merging, two quick questions:

1. Do you suggest we stick with the default browserlist config or we should make IE11 explicit in case the defaults change but we still need to rely on IE11?

2. To _actually_ technically enforce the API linting (i.e. let the linter catch use of unsupported functions such as startsWith), would we need to add something like this? https://github.com/amilajack/eslint-plugin-compat

**Comment by xaverfleer** [comment]
> Do you suggest we stick with the default browserlist config or we should make IE11 explicit in case the defaults change but we still need to rely on IE11?

It is reasonable to stick with the default Browserslist config. As long as IE11 receives browser updates at least once a year, it counts as 'not dead' (hence we support it). Sean Larkin, a Microsoft Edge Developer, shared last week that Microsoft is going to support IE11 "for a long time".

> To actually technically enforce the API linting (i.e. let the linter catch use of unsupported functions such as startsWith), would we need to add something like this? https://github.com/amilajack/eslint-plugin-compat

This sounds like a great idea. Except..
- `eslint-plugin-compat` is not perfect e.g. it [does not support instance methods at the moment](https://github.com/amilajack/eslint-plugin-compat/issues/84).
- all common functions should be polyfilled with [Babel polyfill](https://babeljs.io/docs/en/babel-polyfill) -> we can use the functions in IE11.

Thus I do not see the benefit of adding the tool.

**Comment by xaverfleer** [comment]
Thank you for the fixes 👍. 56e0f36 and 331a321 squashes them with the existing commits to atomic commits.

**Comment by raphaelschaad** [comment]
Awesome. Am I misunderstanding this point from our README then?

- In HTML `<script>` tags JavaScript functions like `startsWith()` [are not supported](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith#Browser_compatibility)

**Comment by xaverfleer** [comment]
You got me. The comment was wrong. b753d63 fixes the comment.

I need to raise my bar. So I will go once again through the whole branch.

**Comment by xaverfleer** [comment]
Thank you for noticing!

**Comment by raphaelschaad** [comment]
As a developer making changes to this code base using a modern browser, what do I have to watch out for given our browser compatibility requirements that isn't caught by linting? How should I be warned?

**Comment by xaverfleer** [comment]
Got it: The comments in "Browser compatibility" were not really about "browser compatibility". decaf09 fixes this.

> How should I be informed?

For now, I think we are okay with code reviews. Eventually we could add webpack to bundle the code (including @babel/polyfill) and test our code with a test runner like Karma that runs tests in actual browsers.

**Comment by raphaelschaad** [comment]
Is the following correct and something I as developer new to this code base have to watch out for?

For JavaScript in `src/`, use latest functions and syntax because we use Babel during building. For JavaScript in `<script>` tags, use latest functions by including the Babel polyfill but **do not use syntax above ES5 (linter won't catch it)** because of support listed in `.browserlists`.

Coming to this code base, should I be informed with a comment in the README?

**Comment by xaverfleer** [comment]
Almost correct:

As a new Developer, I can use any syntax — modern or old — wherever I like — in .js files or in HTML script tags as long as it passes ESLint.

There are only 2 caveats:

1. All HTML files need to contain `<script src="[path]/polyfill.min.js"></script>` (decaf09 adds a comment)
1. ESLint must be run manually (README.md contains a Project setup | Run linting section)

What do you think?

**Comment by raphaelschaad** [comment]
Got it.

#1 and #2 are fine if they're well documented in the README so even I understand them.

So we can use all the latest JS *syntax* (not just functions) in HTML <script> tags? (I assume [this plugin](https://babeljs.io/docs/en/babel-plugin-transform-arrow-functions) is part of the polyfill and does it in the case of arrow functions?)

We do the nifty browserlists integration during the build step [to keep the size of the polyfill small](https://babeljs.io/docs/en/babel-polyfill#size)? (It is referenced in each HTML page now, after all.)

How do you feel about this change to our setup overall?

**Comment by xaverfleer** [comment]
> So we can use all the latest JS syntax (not just functions) in HTML <script> tags?

You got me here. I was so sure, ESLint would throw an error that I did not even try. Curreltly I am in deep dive into ESLint documentation, to be sure to get it right this time.

**Comment by raphaelschaad** [comment]
How are you doing on this? I'm happy to push it over the finish line myself if you won't have any bandwidth in the next 1-2 weeks.

**Comment by xaverfleer** [comment]
I am in deep dive mode to get to a great result. Maybe as overview a see 4 approaches (as of now):
 a) merge it as is (downside: polyfill is very large)
 b) replace polyfill resource with polyfill.io (very powerful, this will load polyfills only for those browsers that need it; downside: it is a 3rd party resource and thus can have an inpact on privacy and reliability)
 c) only polyfill  to browsers that need it similar as described [here](https://hackernoon.com/polyfills-everything-you-ever-wanted-to-know-or-maybe-a-bit-less-7c8de164e423) (downside: this takes some time)
 d) only polyfill functions that are required (downside: apparently this has only little impact on the size)

Also with [eslint-plugin-compat](https://github.com/amilajack/eslint-plugin-compat) we can show lint errors, if unsupported features are used (apparently). This could be very useful.

**Comment by raphaelschaad** [comment]
How about
- [ ] Ensuring this PR addresses #1 and #2 from [the comment above](https://github.com/raphaelschaad/Plus-Web/pull/26#issuecomment-492850142)
- [ ] Doing (d) -- Given we now have browserlists, [integrating during the build step](https://babeljs.io/docs/en/babel-preset-env#browserslist-integration) to [keep the size of the polyfill small](https://babeljs.io/docs/en/babel-polyfill#size) (even though it's not the highest impact according to the [Hacker Noon post](https://hackernoon.com/polyfills-everything-you-ever-wanted-to-know-or-maybe-a-bit-less-7c8de164e423))
- [ ] Adding a GH Issue to do (c) -- write code to conditionally load polyfills only to browsers that need them to reduce size as described in the [Hacker Noon post](https://hackernoon.com/polyfills-everything-you-ever-wanted-

**Comment by xaverfleer** [comment]
I omitted (d) because it is not functional without a bundler like webpack or browserify. Check 1c50c5c out, for the result.

**Comment by xaverfleer** [comment]
todo:
- [ ] reorganize results in atomic commits

**Comment by xaverfleer** [comment]
Requesting a review

**Comment by xaverfleer** [comment]
> Of course, I couldn’t test with, say, IE11, but that shouldn’t matter for this backend functionality.

I agree, this should have no effect.

> In the README, please add [the] sentence [...]

cd278bf adds the sentence

ec4c4f3 makes sure npm packages are up to date

> 🚀

🚀 Thank you for the review! 🚀
